pkg/istio-agent/healthcheck.go

```diff
@@ -0,0 +1,198 @@
+	ExecutableName string
+	Args           []string
+}
+
+// PerformApplicationHealthCheck Performs the application-provided configuration health check.
+// Designed to run async.
+// TODO:
+// 	- Add channel param for quit (better error handling as well)
+// 	- Because there are 3 possible configs, there are 3 possible healthcheck paths, and all of them
+// 		are defined here. Therefore, there is quite a bit of duplicate code in success/fail threshold
+// 		and healthChannel sending. This code should be better.
+// 	- Should the CheckFrequency Delay be a time.Ticker?
+func (w *WorkloadHealthChecker) PerformApplicationHealthCheck(notifyHealthChange chan *discovery.DiscoveryRequest) {
+	// delay before starting probes.
+	time.Sleep(w.config.InitialDelay)
+
+	numSuccess, numFail := 0, 0
+	if w.config.CheckType == HTTPHealthCheck {
+		for {
+			if code, err := httpCheck(w.config.HTTPConfig, w.config.ProbeTimeout); code >= 200 && code <= 299 {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+				}
+			} else {
+				numFail++
+				if numFail == w.config.FailThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{
+						TypeUrl: HealthInfoTypeURL,
+						ErrorDetail: &status.Status{
+							Code:    int32(code),
+							Message: err.Error(),
+						},
+					}
+					numFail = 0
+				}
+			}
+			// should this be a time.Ticker?
+			time.Sleep(w.config.CheckFrequency)
+		}
+	}
+
+	if w.config.CheckType == TCPHealthCheck {
+		for {
+			if err := tcpCheck(w.config.TCPConfig, w.config.ProbeTimeout); err == nil {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+				}
+			} else {
+				numFail++
+				if numFail == w.config.FailThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{
+						TypeUrl: HealthInfoTypeURL,
+						ErrorDetail: &status.Status{
+							Code:    int32(500),
+							Message: err.Error(),
+						},
+					}
+					numFail = 0
+				}
+			}
+		}
+	}
+
+	if w.config.CheckType == ExecHealthCheck {
+		for {
+			if err := execCheck(w.config.ExecConfig); err == nil {
+				numSuccess++
+				if numSuccess == w.config.SuccessThresh {
+					notifyHealthChange <- &discovery.DiscoveryRequest{TypeUrl: HealthInfoTypeURL}
+					numSuccess = 0
+				} else {
+					numFail++
+					if numFail == w.config.FailThresh {
+						notifyHealthChange <- &discovery.DiscoveryRequest{
+							TypeUrl: HealthInfoTypeURL,
+							ErrorDetail: &status.Status{
+								Code:    int32(500),
+								Message: err.Error(),
+							},
+						}
+						numFail = 0
+					}
+				}
+			}
+		}
+	}
+}
+
+// httpCheck performs a http get to a given endpoint with a timeout, and returns
+// the status and error.
+func httpCheck(config HTTPHealthCheckConfig, timeout time.Duration) (int, error) {
+	client := http.Client{
```

Comment:
I think the http.Client object keeps track of connections. In particular, creating a new client will mean we create a new connection for each request. I am not sure if this is what we want - what we really want is whatever kubelet does.

Reply:
btw we also have our own code doing something very similar to the http part which we can probably get best practices from: https://github.com/istio/istio/blob/aa117ca3f7e8e39538ed762cbcefb6ae60e99b2b/pilot/cmd/pilot-agent/status/server.go#L458

https://istio.io/latest/docs/ops/configuration/mesh/app-health-check/

Reply:
Basically its a proxy from kubelet -> agent -> app.

We could possibly share code, but its small anyways

Reply:
I used `http.Client` because it let me set a timeout, which the user provides in the configuration.

It seems https://github.com/istio/istio/blob/aa117ca3f7e8e39538ed762cbcefb6ae60e99b2b/pilot/cmd/pilot-agent/status/server.go#L458 also uses a httpClient, so im not sure of the difference.

is there a particular issue with creating a new connection every time, as long as we close them?